lib/src/main/proto/bio/terra/tanagra/underlay/text.proto

```diff
@@ -0,0 +1,27 @@
+// Messages for information about hierarchical attributes.
+
+syntax = "proto3";
+
+package bio.terra.tanagra.underlay;
+option java_package = "bio.terra.tanagra.proto.underlay";
+option java_multiple_files = true;
+
+import "bio/terra/tanagra/underlay/tables.proto";
+
+// Information required to do a text search for an entity instance.
+message Text {
+  // The name of the entity to search.
+  string entity = 1;
+
+  TextTable text_table = 2;
+}
+
+// The table of information required to perform the text search.
+// There should be a row (lookupTableKey, fullText) for each possible
+// text string to match against per entity.
+message TextTable {
+  // A column that foreign keys to the entity primary key.
+  ColumnId lookup_table_key = 1;
```

Comment:
I don't quite follow why this layer of misdirection is necessary. It seems like this is a design decision around responsibility of the client to construct a query vs how much semantic knowledge is being encoded into the underlay.

Do you have a concrete example to help illustrate this? i.e. showing the tradeoff between
1. the client constructing a filter on their target entity which includes a relationship filter + text match filter, vs
2. the client constructing a filter on their target entity with a text match (underlay encodes the relationship)

Is this because we don't want this relationship  to exist in the entity model?

Reply:
Right now (i.e. with all 3 of these related PRs together, including the workflow one), doing a text search for an entity means doing an exact substring match on a single column in a table that has one row per entity. We could accomplish the same thing by defining a new entity attribute that points to this column. However, I think that approach will not be flexible enough for future changes. e.g.:
- Allowing multiple rows per entity instance in this text search info table (e.g. one row for each synonym or language).
- Allowing multiple columns per entity instance in this text search info table (e.g. text string to search, a relevance ranking).

I think you're right about this being a design choice between how much information the client needs to know vs how much information we encode into the underlay. With this `Text` section of the underlay, the client doesn't need to know how text search is implemented -- they just say "do a text search on entity A". A more client-heavy approach could have been to let the client specify a list of entity attributes to perform the text search on (e.g. text search on name, id and concept_code). That would be a larger burden on the client and also force us to expose all text searchable strings as entity attributes, which may not make much sense (e.g. when we have a list of synonyms, it's not really 1-to-1). I think that could also become a problem once we get to performance tuning because it will be harder to build a good index when the client can specify a variable number of fields to text search on.

RE relationship filter: Relationships require two entities. For text search, it's not clear to me what the second entity would be. To do a text search for conditions, we have the condition entity already, but would need to define a text-search-condition entity? I generally want to think of entities as "objects the UI (user) cares about". So the user can explain what a condition is, but probably not a text-search-condition.

I'm not sure I've answered your questions here, particularly the part about relationships I may be off base to what you were thinking. Please rephrase/ask again if not, and maybe this will be good to talk about in person at stand-up today or in a separate time block if needed.

Reply:
Yes, you answered my question - thanks. I think this makes sense to me as a guiding principle. Follow-up question based on your examples:

Is the cardinality correct here? Is it reasonable to have an "at most one" relationship between an entity, and text search configuration? For example, is it possible that we'd want to support text search on two different logical attributes (or multiple search configurations on the same attribute)? Or if tying the search to a specific attribute is too limiting (as you mention above, maybe you want to search multiple attributes with one operation), should it still be possible to have multiple distinct text search types on the same entity?

This could definitely be a future extension, if there's no foreseeable need for it

Reply:
RE cardinality: You're right we may want multiple configurations of text search in the future. I think the exact use case will influence how we decide to implement it.

e.g. fuzzy text search & exact match text search
For this, I'd expect we could reuse the same index table and just use a different algorithm (e.g. bigquery function) to do the query. So we could continue to restrict it to 1 `Text` block in the underlay per entity.

e.g. english text search & japanese text search
For this, I'd expect we'd have different index tables and would need two pointers to dataset tables per entity. So we'd need to allow >1 `Text` block in the underlay per entity.

As you say, I expect we'll need to revisit this in the future, but I don't see an immediate use case for >1 text search index table per entity.